### ISSUE TITLE
Headers are received in Camel-Case

### ISSUE THREAD

**Issue by jiaming10** [aws/aws-sam-cli#1860]
When running `sam local start-api`, I'm getting headers in Camel-Case, (see **X-Api-Key** and **X-User-Id**), however. the deployed lambda function is receiving the headers in lower-case (**x-api-key** and **x-user-id**).

My local code is working as of 2020-03-09 17:00 GMT+8, and I'm trying to understand why the behaviour is different now.

In `sam local`:
```python
{
  'headers': {
    'X-Api-Key': '<redacted>',
    'X-User-Id': '<redacted>',
    'User-Agent': 'PostmanRuntime/7.22.0',
    'Accept': '*/*',
    'Cache-Control': 'no-cache',
    'Host': 'localhost:3000',
    'Accept-Encoding': 'gzip, deflate, br',
    'Connection': 'keep-alive',
    'X-Forwarded-Proto': 'http',
    'X-Forwarded-Port': '3000'
  },
  'multiValueHeaders': {
    'X-Api-Key': ['<redacted>'],
    'X-User-Id': ['<redacted>'],
    'User-Agent': ['PostmanRuntime/7.22.0'],
    'Accept': ['*/*'],
    'Cache-Control': ['no-cache'],
    'Host': ['localhost:3000'],
    'Accept-Encoding': ['gzip, deflate, br'],
    'Connection': ['keep-alive'],
    'X-Forwarded-Proto': ['http'],
    'X-Forwarded-Port': ['3000']
  }
}
```

In AWS Lambda:
```python
{
  'headers': {
    'Accept': '*/*',
    'Accept-Encoding': 'gzip, deflate, br',
    'Cache-Control': 'no-cache',
    'Host': '<redacted>',
    'User-Agent': 'PostmanRuntime/7.22.0',
    'X-Amzn-Trace-Id': 'Root=1-5e6739e1-d6161c849faa6e9a369f680c',
    'x-api-key': '<redacted>',
    'X-Forwarded-For': '175.143.101.237',
    'X-Forwarded-Port': '443',
    'X-Forwarded-Proto': 'https',
    'x-user-id': '<redacted>'
  },
  'multiValueHeaders': {
    'Accept': ['*/*'],
    'Accept-Encoding': ['gzip, deflate, br'],
    'Cache-Control': ['no-cache'],
    'Host': ['<host>'],
    'User-Agent': ['PostmanRuntime/7.22.0'],
    'X-Amzn-Trace-Id': ['Root=1-5e6739e1-d6161c849faa6e9a369f680c'],
    'x-api-key': ['<redacted>'],
    'X-Forwarded-For': ['<IP address>'],
    'X-Forwarded-Port': ['443'],
    'X-Forwarded-Proto': ['https'],
    'x-user-id': ['<redacted>']
  }
}
```

Snippet of `template.yaml` I'm using.
```yml
Resources:
  BasePythonLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
      LayerName: pythonBaseLayer
      ContentUri: /app/LambdaCodes/LambdaLayers/PythonLayer
      CompatibleRuntimes:
        - python3.7
  VendorFunction:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: vendor-function
      Handler: vendorFunction.lambda_handler
      Runtime: python3.7
      Timeout: 120
      CodeUri: ./VendorFunction/
      Layers:
        - !Ref BasePythonLayer
      Events:
        Vendor:
          Type: Api
          Properties:
            Path: /vendor
            Method: ANY
```

1. OS: Using docker, `python:alpine`
2. `sam --version`: SAM CLI, version 0.40.0

**Comment by jiaming10** [comment]
I still can't figure out why this is happening, so I opted for a workaround instead. The following snippet will inject lowercase headers.
```python
headers = ["X-User-Id","X-Api-Key"]

for header in headers:
    if header in event["headers"]:
        event["headers"][header.lower()] = event["headers"][header]
```

**Comment by denis-ryzhkov** [comment]
More generic workaround:
```
def fix_headers(event):
    """
    Names of custom HTTP headers are provided in:
    - Camel-Case by `sam local start-api`
    - lower-case by the deployed lambda functions

    This workaround adds lower-case names of all HTTP headers

    TODO: Delete it when https://github.com/awslabs/aws-sam-cli/issues/1860 is fixed
    and we use the new version of `aws-sam-cli`
    """
    event["headers"].update(
        {name.lower(): value for name, value in event["headers"].items()}
    )
```

**Comment by jfuss** [comment]
This isn't something we can easily address without re-writing all of the commands built on top of Flask. Kind of relevant here as well: https://github.com/awslabs/aws-sam-cli/issues/621#issuecomment-418373900

Basically, Flask is changing the headers to camel-case and we cannot control that behavior. The fix here is to handle this in your function code. Do note that headers are case-insensitive so handling this in your function code is a reasonable expectation.

Closing this as this is being caused by the library we built upon and the only way out is to re-write (which is a very high cost).

**Comment by denis-ryzhkov** [comment]
This is not a bug for Flask, but it starts being a bug exactly in aws-sam-cli, because `sam local start-api` should be consistent with deployed Lambda functions.

So if this bug is not fixed here, then all multiple apps depending on aws-sam-cli will have to fix it again and again.

Is it that complex to inject `fix_headers(event)` on aws-sam-cli level? It does not look like complete rewrite, let me check...

**Comment by jfuss** [comment]
@denis-ryzhkov How do you know the source headers are suppose to be lower? If we lower case everything and the request had CamelCase then what? We solved your problem but then created the same problem for others. I understand it is a difference and will raise up to the team, but I have looked into this before. 

As I stated, headers are case-insensitive and this can be just as easily handled in the function code.

**Comment by jfuss** [comment]
@denis-ryzhkov Also, let's please be respectful on issues.

**Comment by jfuss** [comment]
Reading the initially report again, we can adjust the known lower case headers to match API Gateway but that does not solve the whole problem here. Might be good enough for now though.

**Comment by denis-ryzhkov** [comment]
Sorry, I didn't want to offend. Could you please point to exact words that you find disrespectful? I'd be grateful to learn how to speak better.

Answering your questions:

> How do you know the source headers are suppose to be lower?

This is not needed, because deployed Lambda functions (or API Gateway) don't preserve the case too, so we just need to be consistent with them.

> If we lower case everything and the request had CamelCase then what?

If we drop original CamelCase header, then we break compatibility, so I proposed to just add lower-case versions, keeping original headers too, for compatibility. It may double the size of headers dict, but it is way less issue, than inconsistency with deployed Lambda.

> this can be just as easily handled in the function code.

Yes, it can be, but it shouldn't be. It should be handled here. I just try to argument this, not to offend.

> we can adjust the known lower case headers to match API Gateway

But this issue is about unknown custom headers, so it's better to add a lower-case copy of all headers to fix this issue.

**Comment by jfuss** [comment]
@denis-ryzhkov 

> This is not needed, because deployed Lambda functions (or API Gateway) don't preserve the case too, so we just need to be consistent with them.

I am slightly confused now. If API Gateway doesn't preserve cases, then what is the issue? If you have to handle all cases because API Gateway doesn't preserve then what we do locally shouldn't matter and this is something you have to handle in your function code anyways. I must be missing something here.

> If we drop original CamelCase header, then we break compatibility, so I proposed to just add lower-case versions, keeping original headers too, for compatibility. It may double the size of headers dict, but it is way less issue, than inconsistency with deployed Lambda.

This doesn't cover everything though. Which might be ok, but the root of the issue here is Flask is Camel-Casing all Headers. Putting them all to lower, still creates problems where the casing is not Camel-Case or all lower. To me, this opens Pandora's box. Do we have to include all headers in all cases to allow this to work? I know I am going to the extreme here but things can increase here. This will also create huge payloads for some functions and make the events not match what API Gateway does, furthering ourselves from the service. 

> Yes, it can be, but it shouldn't be. It should be handled here. I just try to argument this, not to offend.

Why not? If headers are case-insensitive and from your first statement API Gateway doesn't preserve casing, you have to handle this already.

> But this issue is about unknown custom headers, so it's better to add a lower-case copy of all headers to fix this issue.

From the original post: " I'm getting headers in Camel-Case, (see X-Api-Key and X-User-Id), however. the deployed lambda function is receiving the headers in lower-case (x-api-key and x-user-id)." So this is about us camel-casing headers that are lower case, specifically ones we know will be lower. If you look deeper at the differences, there are still Camel-Casing that is in both events. 

Adding lower doesn't solve the root problem here either. Unknown custom headers can be any casing. Adding lower would then be no different than the CamelCases for customers who do not add headers as lower. If you send a header name "MYCUSTOMHEADER" and then could only access it as "mycustomheader" or "Mycustomheader" that is weird. If we had a way to know the casing of the original request, that is what we should be doing. However, Flask doesn't allow that (that I know of) and we would be in the position to re-write to keep the events from not blowing up and match the actually event. So lower casing all headers is and adding them to the event is not something we should be doing here. 

> Sorry, I didn't want to offend. Could you please point to exact words that you find disrespectful? I'd be grateful to learn how to speak better.

"Is it that complex to inject fix_headers(event) on aws-sam-cli level? It does not look like complete rewrite, let me check..." This comes across off as passive aggressive to some degree. Mostly due to the last sentence.

**Comment by denis-ryzhkov** [comment]
Thank you for clarifications, @jfuss , and sorry for the tone, I will be more careful with it.

Now I see that the fix I proposed leads to a new issue: e.g. `event["headers"]["authorization"]` will work in patched `sam local start-api` as a side-effect of adding lower-cased names of all headers. A user may expect this code to work after deploy, but API Gateway will lower-case names of unknown custom headers only, keeping well-known `Authorization` Camel-Cased. The code will be broken, again.

So both current situation and this fix are inconsistent with API Gateway.

I see more options:

1. Get the code API Gateway uses to change the case of some headers, copy it to aws-sam-cli, and keep it in sync when API Gateway decides to change it.

2. As names of the headers should be case-insensitive, follow the `requests` lib in making headers a `CaseInsensitiveDict` as in https://github.com/psf/requests/blob/master/requests/structures.py

While option 2 looks simple and nice, it should be implemented both in aws-sam-cli and in API Gateway to avoid the `event["headers"]["authorization"]` issue above.

However, even if API Gateway may adopt option 2 much later (if ever), at least aws-sam-cli could provide the correct case-insensitive access to headers, like the `requests` lib does. It will not make the situation worse, but will pave a way to the complete fix at both sides.

If you agree, could you please implement option 2, or should I submit a PR for it?

**Comment by jiaming10** [comment]
@jfuss Thanks for the taking the time to respond. I did came across https://github.com/awslabs/aws-sam-cli/issues/621#issuecomment-418373900 when I was researching on the problem. However, it doesn't explain why it `sam local` having a different behaviour after 2020-03-09 17:00 GMT+8. 

> > How do you know the source headers are suppose to be lower?
> 
> This is not needed, because deployed Lambda functions (or API Gateway) don't preserve the case too, so we just need to be consistent with them.

@denis-ryzhkov Actually this is incorrect. As far as I can tell, API Gateway does preserve case. For example if I pass in `x-api-key`, I will receive `x-api-key` at my Lambda code. So I have to agree with @jfuss where lowercasing all the headers won't help as much.

As of now, this is just a minor annoyance to me, but I want to know deeper on why `sam local` having a different behaviour.

**Comment by denis-ryzhkov** [comment]
@jiaming10

> > deployed Lambda functions (or API Gateway) don't preserve the case too, so we just need to be consistent with them.
> 
> @denis-ryzhkov Actually this is incorrect. As far as I can tell, API Gateway does preserve case. For example if I pass in `x-api-key`, I will receive `x-api-key` at my Lambda code.

Let's check:

```
# Request to Lambda deployed to API Gateway:

curl REDACTED.execute-api.us-east-1.amazonaws.com/REDACTED \
    -H "Cache-Control: no-cache" \
    -H "X-Test-Camel-Dashes: 1" \
    -H "x-test-lower-dashes: 2" \
    -H "X-TEST-UPPER-DASHES: 3" \
    -H "X_Test_Camel_Underscores: 4" \
    -H "x_test_lower_underscores: 5" \
    -H "X_TEST_UPPER_UNDERSCORES: 6" \
    -H "Authorization: Bearer 7" \
    --data-binary REDACTED

# CloudWatch Logs of event["headers"]
# https://console.aws.amazon.com/cloudwatch/home#logEventViewer:REDACTED

'x-test-camel-dashes': '1',
'x-test-lower-dashes': '2',
'x-test-upper-dashes': '3',
'x_test_camel_underscores': '4',
'x_test_lower_underscores': '5',
'x_test_upper_underscores': '6'
'Authorization': 'Bearer 7',

'accept': '*/*',
'cache-control': 'no-cache',
'content-type': 'application/x-www-form-urlencoded',
'Host': 'REDACTED',
'User-Agent': 'curl/7.58.0',
'X-Amzn-Trace-Id': 'REDACTED',
'X-Forwarded-For': 'REDACTED',
'X-Forwarded-Port': '443',
'X-Forwarded-Proto': 'https',
```

**This test proves that deployed Lambda functions / API Gateway don't preserve the case of HTTP headers.**

* All explicitly requested custom headers became lower-cased
(that's why your `x-api-key` was "preserved"),
which is unlike Flask and current aws-sam-cli
* In these headers, underscores were preserved, unlike Flask and current aws-sam-cli
* Explicitly requested `Cache-Control` became lower-cased `cache-control`
* Explicitly requested `Authorization` was kept Camel-Cased
* More headers were added by curl and API Gateway implicitly,
case is either lower like `accept` or Camel like `Host`
* Probably, curl sent lower-case `accept`
and API Gateway added Camel-Case `Host`,
as well as Camel-Case `X-Forwarded-For` and others

Such inconsistency both between sam-aws-cli and API Gateway
and also inside of API Gateway itself (`accept` vs `Host`)
adds motivation to make `event["headers"]` a `CaseInsensitiveDict` in both places,
so that headers could be accessed in a case-insensitive way, as they should be:
`event["headers"]["Accept"] == event["headers"]["accept"]`

Do you agree, @jfuss ?

**Comment by denis-ryzhkov** [comment]
Updated workaround:

```python
def fix_headers(event):
    """
    `sam local start-api` from aws-sam-cli is based on Flask,
    which doesn't preserve the case of headers in one way:
    https://github.com/awslabs/aws-sam-cli/issues/1860#issuecomment-628076065

    Lambda is deployed to AWS API Gateway,
    which doesn't preserve the case of headers in another way:
    https://github.com/awslabs/aws-sam-cli/issues/1860#issuecomment-628576313

    Until both aws-sam-cli and API Gateway
    start providing case-insensitive access to headers
    and treating underscores in a uniform way,
    this workaround should be used and underscores should be avoided in headers

    It follows they way how `requests` lib deals with headers:
    https://github.com/psf/requests/blob/master/requests/structures.py

    Usage:
        import json, logging
        from requests.structures import CaseInsensitiveDict
        from traceback import format_exc

        def fix_headers(event): ...  # This workaround

        def endpoint(func):  # Example of decorator to inject this workaround
            def inner(event, context):
                try:
                    fix_headers(event)
                    result = func(event, context)
                    body = json.dumps(result)
                    status = 200

                except Exception:
                    logging.error(format_exc())
                    body = json.dumps({
                        "ok": False,
                        "error": "Internal Serverless Error",
                    })
                    status = 500

                return {
                    "statusCode": status,
                    "headers": {"Content-Type": "application/json"},
                    "body": body,
                }

            return inner

        @endpoint
        def some_lambda_function_handler(event, context):
            assert (
                event["headers"]["Content-Type"] ==
                event["headers"]["content-type"]
            )
            return {"ok": True}
    """
    if not isinstance(event["headers"], CaseInsensitiveDict):
        event["headers"] = CaseInsensitiveDict(event["headers"])
```

**Comment by jfuss** [comment]
@denis-ryzhkov 
> Such inconsistency both between sam-aws-cli and API Gateway
and also inside of API Gateway itself (accept vs Host)
adds motivation to make event["headers"] a CaseInsensitiveDict in both places,
so that headers could be accessed in a case-insensitive way, as they should be:
event["headers"]["Accept"] == event["headers"]["accept"]

CaseInsensitiveDict still doesn't help us because we serialize that dictionary into JSON which is then sent to the function. The event is case-sensitive, so this still leads you to handling the json data in your function in a case-insensitive way. This is what I have been suggesting because Headers are case-insenstive: https://stackoverflow.com/a/5259004. 

If API Gateway documents how they handle casing in Headers, we can create some heuristics to help here but my assumption is that doesn't really help unless we know the raw headers that were sent (which we do not have access to through Flask). 

@jiaming10 
> However, it doesn't explain why it sam local having a different behaviour after 2020-03-09 17:00 GMT+8.

I am not aware of an behavior changing during this time. As far as I remember, this has always been the case locally. 

> As of now, this is just a minor annoyance to me, but I want to know deeper on why sam local having a different behaviour.

SAM CLI has a different behavior because we have to re-implement an AWS Service locally based on docs and behavior we see. When we first moved from Go to Python, we looked at libraries in Python that we could use to help us get an HTTP Service locally without having to build the whole thing from scratch. We ended up choosing Flask (reasons I don't specifically remember). This gets us most of the parity we needed to mimic API Gateway locally. Since we mimic and API Gateway doesn't have a library that we can use to simulate exactly what the cloud does, we are left with these small differences we between the cloud and local. The only way to be 100% (or nearly) would be to re-write API Gateway locally in Python, which is just not feasible.

**Comment by denis-ryzhkov** [comment]
@jfuss 

Thank you for explanation. Looks like local Lambda runner may be not a Python code at all:
https://github.com/awslabs/aws-sam-cli/blob/master/samcli/local/lambdafn/runtime.py#L26

Maybe it is possible to modify this container/JAR or to inject `CaseInsensitiveDict` after loading it from JSON, but it looks too complex now.

So, we will have to use a workaround in each app, as you said. I guess, this issue could be closed.

**Comment by jiaming10** [comment]
@denis-ryzhkov
> **This test proves that deployed Lambda functions / API Gateway don't preserve the case of HTTP headers.**

Thanks for proving me wrong. That is a good reference for me in future.

@jfuss 
> I am not aware of an behavior changing during this time. As far as I remember, this has always been the case locally.

From my end, the behaviour of the function definitely changed. I have tried looking into the docker container's release date, as well as aws-sam-cli, but there are no updates or releases during that period, so I couldn't explain that change. With that said, since the situation that I faced is expected behaviour, I'm more than happy to accept it and move on. 

> > As of now, this is just a minor annoyance to me, but I want to know deeper on why sam local having a different behaviour.

I think my sentence is confusing, I actually meant sam local having a different behaviour after the mentioned date. But thanks for your explanation, now I can understand more on differences between SAM CLI and AWS.

**Comment by sapessi** [comment]
I'd recommend updating the function code to use a [`CaseInsensitiveDict` object just like `requests` does](https://github.com/psf/requests/blob/master/requests/structures.py#L15-L86). In the HTTP protocol, [headers are meant to be case insensitive](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2). Right now it's the `x-api-key` header but in the future the issues could apply to any other header.

**Comment by mastazi** [comment]
@jfuss  said:
>  If we lower case everything and the request had CamelCase then what?

That's fine because it's how API Gateway works. I have submitted PR #3117 to address this.

**Comment by mastazi** [comment]
I have closed #3117 for the reasons explained here by @jfuss https://github.com/aws/aws-sam-cli/issues/3083#issuecomment-887632702
